### DISCUSSION TITLE
Guide: Swap Transaction

### DISCUSSION THREAD

**Discussion by kkwcb** [sol4k/sol4k#61] [Q&A] [answered]
Hello,

Could you please provide an example of how someone would swap tokens?  Now you would use a swap aggregator like Jupiter or Raydium. What I am struggling with is pairing their API with `sol4k`. 

From your example:
```
...
val instruction = TransferInstruction(sender.publicKey, receiver, lamports = 1000)
val transaction = Transaction(blockhash, instruction, feePayer = sender.publicKey)
...
```

I assume I would replace `instruction` with what I get from https://station.jup.ag/docs/apis/swap-api (please see **_"5. Get the serialized transactions to perform the swap"_** section). The name `TransferInstruction` is just throwing me off a bit. Is any transaction just a `TransferInstruction` or would we need to implement `SwapInstruction`?

Thank you!

**Comment by Shpota** [comment, marked as answer]
Hi @itsmelodewyk 

Thank you for opening this discussion. I am sorry you were confused with the APIs.

`TransferInstruction` is meant to transfer SOL from one account to another, generally you cannot use it for other purposes. If you want to include an arbitrary bytes to a transaction you can use `BaseInstruction`. This means you would need to prepare the binary data needed for the instruction yourself. 

However, quickly looking at the API that you shared, I see that they create a _transaction_ for you which you later need to sign. Keep in mind, transactions and instructions are two completely different entities. Every transaction includes one or more instructions. 

Unfortunately, deserialising transactions is not currently implemented in sol4k. You wouldn't be able to complete the first part for the step 6 with sol4k. I am planing to add this functionality in future. You can check [this repository with examples](https://github.com/sol4k/sol4k-examples) to see the main flows that are currently supported.

**Comment by kkwcb** [comment]
Hello @Shpota thank you for the quick reply! I'll take a look at the `BaseInstruction` class. Jupiter's API can generate an instruction as well (under **_"Instructions Instead of Transaction"_** section). Which looks like it might be able to work, however, that would need to be deserialized too but looks manageable. I got some tinkering to do!

**Reply by dovudo** [reply]
So, did you find any solution to swap via sol4k ?

**Reply by 4ig** [reply]
An update for everyone landing on this thread: transaction deserialization has been supported since sol4k `0.5.0`, so this now works end to end. You don't need to build the swap instructions yourself — request a swap transaction from Jupiter's Swap API (or any other aggregator), deserialize it, sign it, and send it:

```kotlin
val connection = Connection(RpcUrl.MAINNNET)
val keypair = Keypair.fromSecretKey(secretKeyBytes)

// swapTransaction is the base64-encoded transaction returned
// by Jupiter's POST /swap endpoint
val transaction = VersionedTransaction.from(swapTransaction)
transaction.sign(keypair)
val signature = connection.sendTransaction(transaction)
```
